Documentation for wan fed via mgw on k8s (#7882)
hashicorp/consul: website/pages/docs/k8s/installation/multi-cluster/kubernetes.mdx

---
layout: docs
page_title: Federation Between Kubernetes Clusters
sidebar_title: Federation Between Kubernetes Clusters
description: >-
  Federating multiple Kubernetes clusters.
---

# Federation Between Kubernetes Clusters

-> **1.8.0+:** This feature is available in Consul versions 1.8.0 and higher

~> This topic requires familiarity with [Mesh Gateways](/docs/connect/mesh_gateway) and [WAN Federation Via Mesh Gateways](/docs/connect/wan-federation-via-mesh-gateways).

-> Looking for a step-by-step guide? Please follow our Learn Guide: [Secure and Route Service Mesh Communication Across Kubernetes](https://learn.hashicorp.com/consul/kubernetes/mesh-gateways).

> This page describes how to federate multiple Kubernetes clusters.

Reviewer: Should we add a use case sentence for federation here, maybe something like "A common scenario for needing federation is to be able to run the same application across multiple Kubernetes clusters that span datacenters or regions" ?

Author: I've listed use cases in the overview page for this multi-cluster section. Do you think I should add it here as well?

```
# Multi-Cluster Federation Overview

In Consul, federation is the act of joining two or more Consul datacenters.
When datacenters are joined, Consul servers in each datacenter can communicate
with one another. This enables the following features:

- Services on all clusters can make calls to each other through Consul Service Mesh.
- [Intentions](/docs/connect/intentions) can be used to enforce cross-cluster rules about which services can communicate.
- [L7 Routing Rules](/docs/connect/l7-traffic-management) can enable multi-cluster failover
  and traffic splitting.
- The Consul UI has a drop-down menu that lets you navigate between datacenters.

```

Author: I added a link to the overview:
```
This page describes how to federate multiple Kubernetes clusters. See [Multi-Cluster Overview](/docs/k8s/installation/multi-cluster/overview)
for more information on use-cases and how it works.
```

## Primary Datacenter

In order to federate clusters, one cluster must be designated the

> primary datacenter. This cluster will be

Reviewer 2: Do you think it's worth a single sentence or so to state that each Kube cluster will be modelled as a separate "Datacenter" regardless of the actual region or zone it's deployed in? Just trying to read this as someone who doesn't know Consul well to start with and the first sentence could be a confusing start?

Reviewer 2: ```suggestion
primary datacenter. This datacenter will be
```
"Cluster" is a bit problematic in Consul usage generally but here it could mean "Kube cluster" which isn't wrong but doesn't make as much sense to me when we are talking specifically about what Consul is doing within this cluster and not Kube itself. Not a big deal though if others disagree!

Author: > Do you think it's worth a single sentence or so to state that each Kube cluster will be modelled as a separate "Datacenter" regardless of the actual region or zone it's deployed in? Just trying to read this as someone who doesn't know Consul well to start with and the first sentence could be a confusing start?

Added "Consul treats each Kubernetes cluster as a separate Consul datacenter."

> "Cluster" is a bit problematic in Consul usage generally but here it could mean "Kube cluster" which isn't wrong but doesn't make as much sense to me when we are talking specifically about what Consul is doing within this cluster and not Kube itself. Not a big deal though if others disagree!

Totally agree, I've tried to ride the line where if I'm talking about Kubernetes I used cluster and if I'm talking about Consul-specific things I used datacenter. In this case datacenter is the right word in this context and I just missed it.

> responsible for creating the certificate authority that will be used to encrypt
> Connect traffic. It also handles validating global ACL tokens. All other clusters

Reviewer 2: ```suggestion
responsible for creating the certificate authority that signs the TLS certificates Connect 
uses to encrypt and authorize traffic. It also handles validating global ACL tokens. All other clusters
```

If you haven't installed Consul on your cluster, continue reading below. If you've
already installed Consul on a cluster and want to upgrade it to
support federation, see [Upgrading An Existing Cluster](#upgrading-an-existing-cluster).

You will need to use the following `config.yaml` file for your primary cluster,
with the possible modifications listed below.

```yaml
global:
  name: consul
  image: consul:1.8.0-beta1
  datacenter: dc1

  # TLS configures whether Consul components use TLS.
  tls:
    # TLS must be enabled for federation in Kubernetes.
    enabled: true

  federation:
    enabled: true
    # This will cause a Kubernetes secret to be created that
    # can be imported by secondary datacenters to configure them
    # for federation.
    createFederationSecret: true

  acls:
    manageSystemACLs: true
    # If ACLs are enabled, we must create a token for secondary
    # datacenters to replicate ACLs.
    createReplicationToken: true

  # Gossip encryption secures the protocol Consul uses to quickly
  # discover new nodes and detect failure.
  gossipEncryption:
    secretName: consul-gossip-encryption-key
    secretKey: key

connectInject:

>   # Consul connect service mesh must be enabled for federation.

Reviewer 2: ```suggestion
  # Consul Connect service mesh must be enabled for federation.
```

~> **Secret doesn't exist?** If you haven't set `global.name` to `consul` then the name of the secret will
be your Helm release name suffixed with `-consul-federation` e.g. `helm-release-consul-federation`.

Now you're ready to import the secret into your secondary cluster(s).

Switch `kubectl` context to your secondary Kubernetes cluster. In this example
our context for our secondary cluster is `dc2`:

```sh
$ kubectl config use-context dc2
Switched to context "dc2".
```

And import the secret:

```sh
$ kubectl apply -f consul-federation-secret.yaml
secret/consul-federation configured
```

#### Federation Secret Contents

The automatically generated federation secret contains:

- **Server certificate authority certificate** - This is the certificate authority
  used to sign Consul server-to-server communication. This is required by secondary
  clusters because they must communicate with the Consul servers in the primary cluster.
- **Server certificate authority key** - This is the signing key for the server certificate
  authority. This is required by secondary clusters because they need to create
  server certificates for each Consul server using the same certificate authority

>   as the primary.

Reviewer 2: It feels like we should have a security warning box somewhere in this section to make it super clear that access to this secret is equivalent to full-admin privs on the whole Consul install since the TLS cert key effectively allows any attacker to create a server cert that will be accepted by other primary servers who will then replicate all state including ACL tokens and CA keys to the attacker.

Reviewer 2: Similar goes for the gossip token in terms of being able to join the cluster.

Reviewer 2: We typically use the `!> **Security note:**` prefix for these btw.

Author: done, thanks I missed that.

1. The default mesh gateway configuration
   creates a Kubernetes Load Balancer service. If you wish to customize the
   mesh gateway, for example using a Node Port service or a custom DNS entry,
   see the [Helm reference](/docs/k8s/helm#v-meshgateway) for that setting.

With your `config.yaml` ready to go, follow our [Installation Guide](/docs/k8s/installation/overview)
to install Consul on your secondary cluster(s).

## Verifying Federation

To verify that both datacenters are federated, run the
`consul members -wan` command on one of the Consul server pods:

```sh
$ kubectl exec statefulset/consul-server -- consul members -wan
Node                 Address              Status  Type    Build     Protocol  DC   Segment
consul-server-0.dc1  10.32.4.216:8302     alive   server  1.7.0dev  2         dc1  <all>
consul-server-0.dc2  192.168.2.173:8302   alive   server  1.7.0dev  2         dc2  <all>
consul-server-1.dc1  10.32.5.161:8302     alive   server  1.7.0dev  2         dc1  <all>
consul-server-1.dc2  192.168.88.64:8302   alive   server  1.7.0dev  2         dc2  <all>
consul-server-2.dc1  10.32.1.175:8302     alive   server  1.7.0dev  2         dc1  <all>

> consul-server-2.dc2  192.168.35.174:8302  alive   server  1.7.0dev  2         dc2  <all>

Reviewer 2: no big deal but is it worth "fixing" the `Build` column to be 1.8.0-beta?

Author: 🍺

context is `dc1` and we're querying for the list of services in `dc2`:

```sh
$ kubectl exec statefulset/consul-server -- consul catalog services -datacenter dc2
consul
mesh-gateway
```

You can switch kubectl contexts and run the same command in `dc2` with the flag
`-datacenter dc1` to ensure `dc2` can communicate with `dc1`.

### Consul UI

We can also use the Consul UI to verify federation.
See [Viewing the Consul UI](docs/k8s/installation/overview#viewing-the-consul-ui)
for instructions on how to view the UI.

~> NOTE: If ACLs are enabled, your kubectl context must be in the primary datacenter
to retrieve the bootstrap token mentioned in the UI documentation.

With the UI open, you'll be able to switch between datacenters via the dropdown
in the top left:

![Consul Datacenter Dropdown](/img/consul-datacenter-dropdown.png 'Consul Datacenter Dropdown')

## Next Steps

With your Kubernetes clusters federated, try out using Consul service mesh to
route between services deployed on each cluster by following our Learn Guide: [Secure and Route Service Mesh Communication Across Kubernetes](https://learn.hashicorp.com/consul/kubernetes/mesh-gateways#deploy-microservices).

> You can also read our in-depth documentation on [Consul Service Mesh In Kubernetes](/docs/k8s/connect).

Reviewer 2: This are fantastic 🎉 Great job!
